### PR TITLE
Read fluor data

### LINKED ISSUE #606: Support PSII data from CropReporter
We wish to add the functionality to support CropReporter PSII data. Short term goals include the ability to read in raw image data files into a PlantCV workflow, with long term goals of extended analysis capabilities into a fluorescence subpackage. 

**Proposed method**
Dump in raw data, reshape, and possibly filter captured arrays to keep those needed. 
Minimal method for reading the data: 
```
from plantcv import plantcv as pcv
import numpy as np
dat = "image.DAT"
ImageCols = 1296
ImageRows = 966
y = ImageRows
x = ImageCols
raw_data = np.fromfile(dat, np.uint16, -1)
array_data = raw_data.reshape(y, int(len(raw_data)/(y*x)), x).transpose((0, 2, 1))
```

**Describe alternatives you've considered**

**Additional context**
* Crop reporter outputs a DAT file and an INI file
* INI file is sort of like the ENVI (hyperspectral) metadata file, and includes image dimensions 
* In array_data each item in the z-axis/channel dimension is a separate image
* The INI file should help us figure out what each image in the set is

### LINKED ISSUE #613: I found the broken link of the sample code about "analyze_stem".
**Describe the bug**
This's the broken link about "analyze_stem".
https://github.com/danforthcenter/plantcv/blob/master/plantcv/plantcv/analyze_stem.py

**To Reproduce**
Steps to reproduce the behavior (including sample data and code really help):
1.  Go to "https://github.com/danforthcenter/plantcv/blob/master/docs/analyze_stem.md"
2.  Click to the link of "Source Code: Here".
3.  See error

**Expected behavior**
It should see the sample code of "analyze_stem.py".

**Local environment (please complete the following information):**
 - OS: All
 - Environment: Web
 - PlantCV Version: All

**Additional context**
None

### PR DESCRIPTION
**Pull request by HaleySchuhl**
**Describe your changes**
New function to read in fluorescence image files. Will return `fdark`, `fmin`, and `fmax` to be used downstream in existing FVFM analysis functions. 

**Type of update**
Is this a:
* New feature or feature enhancement

**Associated issues**
- closes #606 
- closes #613 

**Additional context**
- [x] add testing 
- [x] finish documentation page
- [x] update psII tutorial 
- [x] test function with downstream analysis

### CODE REVIEW

**Comment by HaleySchuhl** [comment]
Replaced the PSII tutorial with an updated workflow for Fv/Fm analysis.

File: docs/photosynthesis_read_dat.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    - filename - image file to be read (possibly including a path)
+    
+- **Context:**
+    - Reads in file to be processed and does so using the metadata contained within a corresponding .INF file
+- **Notes:**
+    - This function assumes a specific pattern between .DAT image files and their corresponding .INF file. 
+    We assume that for every image file `xx_PSD_xxx.DAT` there will be a corresponding metadata file with the same path
+    named `xx_HDR_xxx.INF` 
+- **Example use:**
+    - [Use In PSII Tutorial](psII_tutorial.md) ****Needs updating****
```
Comment (nfahlgren) [inline docs/photosynthesis_read_dat.md:19]: "Needs updating" note still relevant?

Comment (HaleySchuhl) [reply docs/photosynthesis_read_dat.md:19]: Yes. Once we do another version release, I'll be able to open a PR for the updated tutorial. I've got a branch started right now with the new data and workflow!

Comment (HaleySchuhl) [reply docs/photosynthesis_read_dat.md:19]: Oh wait, Binder needs updating now. This doc page is actually updated now!

File: docs/vis_tutorial.md [hunk 2]
```diff
@@ -490,7 +490,7 @@ In the terminal:
 Python script: 
 
 ```python
-# !/usr/bin/env python
```
Comment (nfahlgren) [inline docs/vis_tutorial.md:493]: It's better if workflow scripts use `#!/usr/bin/env python`. This figures out where Python is in your environment, versus a hardcoded path to `/usr/bin/python`.

File: mkdocs.yml [hunk 3]
```diff
@@ -114,6 +114,8 @@ nav:
       - 'Opening': opening.md
       - 'Output Mask': output_mask.md
       - 'Outputs Class': outputs.md
+      - 'Photosynthesis Functions':
+          - 'Read DAT': photosynthesis_read_dat.md
```
Comment (nfahlgren) [inline mkdocs.yml:118]: We should move the "Analyze PSII" page into this group too.

File: plantcv/plantcv/__init__.py [hunk 4]
```diff
@@ -191,7 +191,7 @@ def __init__(self, array_data, max_wavelength, min_wavelength, max_value, min_va
 from plantcv.plantcv.analyze_bound_vertical import analyze_bound_vertical
 from plantcv.plantcv.analyze_color import analyze_color
 from plantcv.plantcv.analyze_nir_intensity import analyze_nir_intensity
-from plantcv.plantcv.fluor_fvfm import fluor_fvfm
+from plantcv.plantcv.photosynthesis.analyze_fvfm import analyze_fvfm
```
Comment (nfahlgren) [inline plantcv/plantcv/__init__.py:194]:
```suggestion
```

File: plantcv/plantcv/__init__.py [hunk 5]
```diff
@@ -238,14 +239,14 @@ def __init__(self, array_data, max_wavelength, min_wavelength, max_value, min_va
            'rgb2gray', 'median_blur', 'fill', 'invert', 'logical_and', 'logical_or', 'logical_xor',
            'find_objects', 'roi_objects', 'object_composition', 'analyze_object', 'morphology',
            'analyze_bound_horizontal', 'analyze_bound_vertical', 'analyze_color', 'analyze_nir_intensity',
-           'fluor_fvfm', 'print_results', 'resize', 'flip', 'crop_position_mask', 'get_nir', 'report_size_marker_area',
+           'analyze_fvfm', 'print_results', 'resize', 'flip', 'crop_position_mask', 'get_nir', 'report_size_marker_area',
```
Comment (nfahlgren) [inline plantcv/plantcv/__init__.py:242]:
```suggestion
           'print_results', 'resize', 'flip', 'crop_position_mask', 'get_nir', 'report_size_marker_area',
```

File: plantcv/plantcv/photosynthesis/__init__.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+from plantcv.plantcv.photosynthesis.read_dat import read_dat
+
```
Comment (nfahlgren) [inline plantcv/plantcv/photosynthesis/__init__.py:2]:
```suggestion
from plantcv.plantcv.photosynthesis.analyze_fvfm import analyze_fvfm
```

File: plantcv/plantcv/photosynthesis/__init__.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
...
+
+__all__ = ["read_dat"]
```
Comment (nfahlgren) [inline plantcv/plantcv/photosynthesis/__init__.py:4]:
```suggestion
__all__ = ["read_dat", "analyze_fvfm"]
```

File: docs/psII_tutorial.md [hunk 8]
```diff
@@ -8,19 +8,17 @@ The debug has three modes: either None, 'plot', or 'print'. If set to
 the images plot to the screen.
 This allows users to visualize and optimize each step on individual test images and small test sets before workflows are deployed over whole datasets.
 
-PSII images (3 in a set; F0, Fmin, and Fmax) are captured directly following a saturating fluorescence pulse 
-(red light; 630 nm). These three PSII images can be used to calculate Fv/Fm (efficiency of photosystem II) 
-for each pixel of the plant. Unfortunately, our PSII imaging cabinet has a design flaw when capturing images 
-of plants with vertical architecture. You can read more about how we validated this flaw using our PSII 
-analysis workflows in the [PlantCV Paper](http://dx.doi.org/10.1016/j.molp.2015.06.005). 
+PSII images (3 in a set; F0, Fmin, and Fmax) are captured directly following a saturating fluorescence pulse
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:11]:
```suggestion
PSII images (3 in a set; Fdark, Fmin (F0), and Fmax) are captured directly following a saturating fluorescence pulse
```

File: docs/psII_tutorial.md [hunk 9]
```diff
@@ -8,19 +8,17 @@ The debug has three modes: either None, 'plot', or 'print'. If set to
...
+(red light; 630 nm). These three PSII images can be used to calculate Fv/Fm (efficiency of photosystem II)
+for each pixel of the plant. Unfortunately, our PSII imaging cabinet has a design flaw when capturing images
+of plants with vertical architecture. You can read more about how we validated this flaw using our PSII
+analysis workflows in the [PlantCV Paper](http://dx.doi.org/10.1016/j.molp.2015.06.005).
 However, the workflows to analyze PSII images are functional and a sample workflow is outlined below.
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:16]:
```suggestion
for each pixel of the plant. A workflow to analyze PSII images is outlined below.  
```

File: docs/psII_tutorial.md [hunk 10]
```diff
@@ -32,23 +30,23 @@ To run a PSII workflow over a single PSII image set (3 images) there are 4 requi
 
 1.  **Image 1:** F0 (a.k.a Fdark/null) image.
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:31]:
```suggestion
1.  **Image 1:** Fdark (null) image.
```

File: docs/psII_tutorial.md [hunk 11]
```diff
@@ -32,23 +30,23 @@ To run a PSII workflow over a single PSII image set (3 images) there are 4 requi
...
-(for PSII images we use a premade mask to remove the screws from the image). 
-Make sure the input is the same size as your image or you will have problems.  
 
 Sample command to run a workflow on a single PSII image set:  
 
 * Always test workflows (preferably with -D flag for debug mode) before running over a full image set.
 
 ```
-./workflowname.py -i /home/user/images/testimg.png -o /home/user/output-images -D 'print'
+./workflowname.py -i /home/user/images/PSII_PSD_testimg_22_rep6.DAT.DAT -o /home/user/output-images -D 'print'
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:49]:
```suggestion
./workflowname.py -i /home/user/images/PSII_PSD_testimg_22_rep6.DAT -o /home/user/output-images -D 'print'
```

File: docs/psII_tutorial.md [hunk 12]
```diff
@@ -355,7 +259,7 @@ To deploy a workflow over a full image set please see tutorial on [workflow para
 In the terminal:
 
 ```
-./workflowname.py -i /home/user/images/testimg.png -o /home/user/output-images -D 'print'
+./workflowname.py -i /home/user/images/PSII_PSD_testimg_22_rep6.DAT.DAT -o /home/user/output-images -D 'print'
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:262]:
```suggestion
./workflowname.py -i /home/user/images/PSII_PSD_testimg_22_rep6.DAT -o /home/user/output-images -D 'print'
```

File: docs/psII_tutorial.md [hunk 13]
```diff
@@ -58,295 +56,201 @@ Workflows start by importing necessary packages, and by defining user inputs.
...
-    # Store the two images
+    #   fdark           - Grayscale image
+    #   fmin            - Grayscale image
+    #   fmax            - Grayscale image
+    #   mask            - Binary mask of selected contours
+    #   bins            - Number of grayscale bins (0-256 for 8-bit img, 0-65536 for 16-bit). Default bins = 256
+    fvfm_images = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=cleaned_mask, bins=256)
+
+    # Store the two fv_fm images
     fv_img = fvfm_images[0]
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:216]:
```suggestion
    fvfm_img = fvfm_images[0]
```

File: docs/psII_tutorial.md [hunk 14]
```diff
@@ -58,295 +56,201 @@ Workflows start by importing necessary packages, and by defining user inputs.
...
+    #     obj           - Single or grouped contour object (optional), if provided the pseudocolored image gets cropped down to the region of interest.
+    #     mask          - Binary mask (optional)
+    #     background    - Background color/type. Options are "image" (gray_img), "white", or "black". A mask must be supplied.
+    #     cmap          - Colormap (https://matplotlib.org/tutorials/colors/colormaps.html)
+    #     min_value     - Minimum value for range of interest
+    #     max_value     - Maximum value for range of interest
+    #     dpi           - Dots per inch for image if printed out (optional, if dpi=None then the default is set to 100 dpi).
+    #     axes          - If False then the title, x-axis, and y-axis won't be displayed (default axes=True).
+    #     colorbar      - If False then the colorbar won't be displayed (default colorbar=True)
+    pseudocolored_img = pcv.visualize.pseudocolor(gray_img=fv_img, mask=cleaned_mask, cmap='jet')
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:232]:
```suggestion
    pseudocolored_img = pcv.visualize.pseudocolor(gray_img=fvfm_img, mask=cleaned_mask, cmap='jet')
```

File: docs/psII_tutorial.md [hunk 15]
```diff
@@ -385,74 +285,45 @@ def options():
...
+    obj, masked = pcv.object_composition(img=cleaned_mask, contours=id_objects, hierarchy=obj_hierarchy)
+
+     # Find shape properties
+    shape_img = pcv.analyze_object(img=fmax, obj=obj, mask=cleaned_mask)
+
+    # Analyze fv/fm fluorescence properties
+    fvfm_images = pcv.pcv.photosynthesis.analyze_fvfm((fdark=fdark, fmin=fmin, fmax=fmax, mask=cleaned_mask, bins=256)
+
+    # Store the two fv_fm images
     fv_img = fvfm_images[0]
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:317]:
```suggestion
    fvfm_img = fvfm_images[0]
```

File: docs/psII_tutorial.md [hunk 16]
```diff
@@ -385,74 +285,45 @@ def options():
...
     fvfm_hist = fvfm_images[1]
 
     # Pseudocolor the Fv/Fm grayscale image that is calculated inside the fluor_fvfm function
-    pseudocolored_img = pcv.visualize.pseudocolor(gray_img=fv_img, mask=kept_mask, cmap='jet')
+    pseudocolored_img = pcv.visualize.pseudocolor(gray_img=fv_img, mask=cleaned_mask, cmap='jet')
```
Comment (nfahlgren) [inline docs/psII_tutorial.md:321]:
```suggestion
    pseudocolored_img = pcv.visualize.pseudocolor(gray_img=fvfm_img, mask=cleaned_mask, cmap='jet')
```

File: tests/tests.py [hunk 17]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    pcv.params.debug_outdir = cache_dir
+    filename = os.path.join(cache_dir, 'plantcv_fvfm_hist.png')
+    # Read in test data
+    fdark = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FDARK), -1)
+    fmin = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMIN), -1)
+    fmax = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMAX), -1)
+    fmask = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMASK), -1)
+    # Test with debug = "print"
+    pcv.params.debug = "print"
+    _ = pcv.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```
Comment (nfahlgren) [inline tests/tests.py:4395]:
```suggestion
    _ = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```

File: tests/tests.py [hunk 18]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    analysis_images = pcv.analyze_fvfm(fdark=fdark + 3000, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```
Comment (nfahlgren) [inline tests/tests.py:4396]:
```suggestion
    analysis_images = pcv.photosynthesis.analyze_fvfm(fdark=fdark + 3000, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```

File: tests/tests.py [hunk 19]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    # Test under updated print and plot function
+    hist_img = analysis_images[1]
+    pcv.print_image(hist_img, filename)
+    pcv.plot_image(hist_img)
+    # Test with debug = "plot"
+    pcv.params.debug = "plot"
+    _ = pcv.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```
Comment (nfahlgren) [inline tests/tests.py:4403]:
```suggestion
    _ = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```

File: tests/tests.py [hunk 20]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    # Test with debug = None
+    pcv.params.debug = None
+    fvfm_images = pcv.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```
Comment (nfahlgren) [inline tests/tests.py:4406]:
```suggestion
    fvfm_images = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```

File: tests/tests.py [hunk 21]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    assert len(fvfm_images) != 0
+
+
+def test_plantcv_fluor_fvfm_bad_input():
+    fdark = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_COLOR))
+    fmin = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMIN), -1)
+    fmax = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMAX), -1)
+    fmask = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMASK), -1)
+    with pytest.raises(RuntimeError):
+        _ = pcv.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```
Comment (nfahlgren) [inline tests/tests.py:4418]:
```suggestion
        _ = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```

**Review by nfahlgren — CHANGES_REQUESTED** [review]
Thanks @HaleySchuhl and @annacasto!

One suggestion/comment not reviewed above: When we generate the pseudocolored Fv/Fm images I think we should set `min_value` to 0 and `max_value` to 1 instead of rescaling the data. Folks would be familiar with that as the typical Fv/Fm range.

File: docs/vis_tutorial.md (hunk 2 above)
Comment (HaleySchuhl) [reply docs/vis_tutorial.md:493]: I wasn't sure, so I went with the more frequently used option. But I've switched that line in all tutorials to use `#!/usr/bin/env python`

File: plantcv/plantcv/photosynthesis/read_dat.py [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    while i < 21:
+        frame = img_cube[:, :, [i]]
+        frame_raw = np.transpose(np.transpose(frame)[0])
+        pixel_sum = np.sum(frame_raw)
+        if pixel_sum > max_sum:
+            max_sum = pixel_sum
+            max_index = i
+        i += 1
+    fmax = img_cube[:, :, [max_index]]
+    fmax = np.transpose(np.transpose(fmax)[0])  # Reshape frame from (x,y,1) to (x,y)
```
Comment (nfahlgren) [inline plantcv/plantcv/photosynthesis/read_dat.py:79]:
```suggestion
    frame_sums = []
    for i in range(img_cube.shape[2]):
        frame_sums.append(np.sum(img_cube[:, :, i]))
    fmax = img_cube[:, :, np.argmax(frame_sums)]
```

How about this?

File: tests/tests.py [hunk 23]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    hist_img = analysis_images[1]
+    pcv.print_image(hist_img, filename)
+    pcv.plot_image(hist_img)
+    # Test with debug = "plot"
+    pcv.params.debug = "plot"
+    _ = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
+    # Test with debug = None
+    pcv.params.debug = None
+    fvfm_images = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
+    pcv.print_results(os.path.join(cache_dir, "results.txt"))
```
Comment (nfahlgren) [inline tests/tests.py:4407]: Results are saved but not used, can we remove this?

File: tests/tests.py [hunk 24]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    filename = os.path.join(cache_dir, 'plantcv_fvfm_hist.png')
+    # Read in test data
+    fdark = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FDARK), -1)
+    fmin = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMIN), -1)
+    fmax = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMAX), -1)
+    fmask = cv2.imread(os.path.join(TEST_DATA, TEST_INPUT_FMASK), -1)
+    # Test with debug = "print"
+    pcv.params.debug = "print"
+    _ = pcv.photosynthesis.analyze_fvfm(fdark=fdark, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
+    analysis_images = pcv.photosynthesis.analyze_fvfm(fdark=fdark + 3000, fmin=fmin, fmax=fmax, mask=fmask, bins=1000)
```
Comment (nfahlgren) [inline tests/tests.py:4396]: Do we need to run the function both ways? I think if so they should be different tests?

File: tests/tests.py [hunk 25]
```diff
@@ -4401,6 +4364,60 @@ def test_plantcv_hyperspectral_inverse_covariance():
...
+    # Test under updated print and plot function
+    hist_img = analysis_images[1]
+    pcv.print_image(hist_img, filename)
+    pcv.plot_image(hist_img)
```
Comment (nfahlgren) [inline tests/tests.py:4400]: These are not used in testing, could be removed?

**Review by nfahlgren — CHANGES_REQUESTED** [review]
Thanks @HaleySchuhl, here's a few more suggestions!

**Comment by dschneiderch** [comment]
I'd suggest renaming read_dat because that's a pretty generic file extension but the reader is very specific to Crop Reporter with a .ini file. `read_cropreporter`? `read_ini`? There are other data formats for chl fluorescence data.
These PSII data can get pretty large with many frames similar to the hyperspectral data.  why not build a new class and use `read_image`?

**Comment by HaleySchuhl** [comment]
@dschneiderch I agree, we could use a more specific name than `pcv.photosynthesis.read_dat`, and it's definitely intended that eventually this function will be integrated into `pcv.readimage`. The reason we haven't utilized classes and integrated this function into `readimage` is because we want to wrap up v3 as quickly as possible, and then restructure ALL image types into convenient classes. 

I'm good with the renaming to `pcv.photosynthesis.read_crop_reporter` for the PR though! Thoughts @nfahlgren ?

File: tests/tests.py (hunk 23 above)
Comment (HaleySchuhl) [reply tests/tests.py:4407]: I think this was added in to make sure all observations in the Outputs class were able to successfully write out to json format. We used to have datatype errors thrown with some of the data getting stored by different analysis functions. This should probably be it's own test though in each case.

File: tests/tests.py (hunk 24 above)
Comment (HaleySchuhl) [reply tests/tests.py:4396]: Yes, I believe these two cases aim to hit all of the code in this part of `analyze_fvfm`:

```
    if np.amax(fdark_mask) > 2000:
        qc_fdark = False
    else:
        qc_fdark = True
```

I will separate these into two separate tests!

**Comment by nfahlgren** [comment]
Sounds good! How about pcv.photosynthesis.read_cropreporter (minus the second underscore)? I think that's how Phenovation names it.

File: tests/tests.py (hunk 23 above)
Comment (HaleySchuhl) [reply tests/tests.py]: done!

**Review by nfahlgren — APPROVED** [review]

**Pull request merged**